core/src/main/java/jenkins/security/UserDetailsCache.java

```diff
@@ -0,0 +1,190 @@
+import org.springframework.dao.DataAccessException;
+
+import javax.annotation.CheckForNull;
+import javax.annotation.Nonnull;
+import java.util.concurrent.Callable;
+import java.util.concurrent.ExecutionException;
+import java.util.concurrent.TimeUnit;
+
+import static com.google.common.cache.CacheBuilder.newBuilder;
+
+/**
+ * Cache layer for {@link org.acegisecurity.userdetails.UserDetails} lookup.
+ *
+ * @since TODO
+ */
+@Extension
+@Restricted(NoExternalUse.class) //TODO Keep for LTS, Remove when in weekly
+public final class UserDetailsCache {
+
+    private static final String SYS_PROP_NAME = UserDetailsCache.class.getName() + ".EXPIRE_AFTER_WRITE_SEC";
+    /**
+     * Nr of seconds before a value expires after being cached, note full GC will also clear the cache.
+     * Should be able to set this value in script and then reload from disk to change in runtime.
+     */
+    private static /*not final*/ Integer EXPIRE_AFTER_WRITE_SEC = SystemProperties.getInteger(SYS_PROP_NAME, (int)TimeUnit.MINUTES.toSeconds(2));
+    private final Cache<String, UserDetails> detailsCache;
+    private final Cache<String, Boolean> existanceCache;
+
+    /**
+     * Constructor intended to be instantiated by Jenkins only.
+     */
+    @Restricted(NoExternalUse.class)
+    public UserDetailsCache() {
+        if (EXPIRE_AFTER_WRITE_SEC == null || EXPIRE_AFTER_WRITE_SEC <= 0) {
+            //just in case someone is trying to trick us
+            EXPIRE_AFTER_WRITE_SEC = SystemProperties.getInteger(SYS_PROP_NAME, (int)TimeUnit.MINUTES.toSeconds(2));
+            if (EXPIRE_AFTER_WRITE_SEC <= 0) {
+                //The property could also be set to a negative value
+                EXPIRE_AFTER_WRITE_SEC = (int)TimeUnit.MINUTES.toSeconds(2);
+            }
+        }
+        detailsCache = newBuilder().softValues().expireAfterWrite(EXPIRE_AFTER_WRITE_SEC, TimeUnit.SECONDS).build();
+        existanceCache = newBuilder().softValues().expireAfterWrite(EXPIRE_AFTER_WRITE_SEC, TimeUnit.SECONDS).build();
+    }
+
+    /**
+     * The singleton instance registered in Jenkins.
+     * @return the cache
+     */
+    public static UserDetailsCache get() {
+        return ExtensionList.lookup(UserDetailsCache.class).get(UserDetailsCache.class);
+    }
+
+    /**
+     * Gets the cached UserDetails for the given username.
+     * Similar to {@link #loadUserByUsername(String)} except it doesn't perform the actual lookup if there is a cache miss.
+     *
+     * @param idOrFullName the username
+     *
+     * @return {@code null} if the cache doesn't contain any data for the key or the user details cached for the key.
+     * @throws UsernameNotFoundException if a previous lookup resulted in the same
+     */
+    @CheckForNull
+    public UserDetails getCached(String idOrFullName) throws UsernameNotFoundException {
+        Boolean exists = existanceCache.getIfPresent(idOrFullName);
+        if (exists != null && !exists) {
```

Comment:
If `exists` is null it means we have no information if the user exists or not, the very definition of `null` ;)

If it has a value, and it is false; then we can follow the spec and throw the exception telling the caller that the username definitely does not exist.
Otherwise we just check the details cache and it may or may not return `null`.